app/models/blueprint.rb

```diff
@@ -0,0 +1,51 @@
+class Blueprint < ApplicationRecord
+  has_many :service_templates
+  private  :service_templates, :service_templates=
+
+  # the top of the service_templates, a bundle that contains child items
+  def bundle
+    service_templates.find { |st| st.parent_services.blank? }
+  end
+
+  # copy a service template and link its blueprint
+  def copy_service_template(service_template, new_attributes = {})
+    self.class.transaction do
+      service_template.dup.tap do |new_template|
+        duplicate_resource_actions(service_template, new_template)
+        duplicate_service_resources(service_template, new_template)
+        update_copied_service_template(new_template, new_attributes)
+      end
+    end
+  end
+
+  private
+
+  def duplicate_resource_actions(old_template, new_template)
+    new_template.resource_actions = old_template.resource_actions.collect(&:dup)
+  end
+
+  def duplicate_service_resources(old_template, new_template)
+    new_template.service_resources = old_template.service_resources.collect do |sr|
+      sr.dup.tap do |new_sr|
+        resource = sr.resource
+        if resource.kind_of?(ServiceTemplate)
+          new_sr.resource = copy_service_template(resource)
+        elsif !(resource.respond_to?(:ems_ref) || resource.kind_of?(ExtManagementSystem))
```

Comment:
Suggest moving this condition into a method with a descriptive name like `provider_resource?`.  That will reduce the need for the comment and be helpful if the logic gets more complex in the future.